07.ls_object/lib/ls_command.rb

```diff
@@ -0,0 +1,102 @@
+    options = parse_options
+    ls_files = bring_ls_files(options)
+    if options[:long]
+      show_files_by_long_format(ls_files)
+    else
+      show_files_by_default_format(ls_files)
+    end
+  end
+
+  private
+
+  def parse_options
+    options = { all: false, reverse: false, long: false }
+
+    OptionParser.new do |opts|
+      opts.on('-a') { options[:all] = true }
+      opts.on('-r') { options[:reverse] = true }
+      opts.on('-l') { options[:long] = true }
+    end.parse!
+
+    options
+  end
+
+  def bring_ls_files(options)
+    target_directory = Pathname(File.expand_path(ARGV[0] || '.'))
+    files = Dir.foreach(target_directory)
+    file_paths = files.map { |file| target_directory.join(file) }
+    ls_files = file_paths.map { |file_path| LsFile.new(file_path, target_directory, options) }
+    filtered_ls_files = options[:all] ? ls_files : ls_files.reject(&:hidden?)
+    options[:reverse] ? filtered_ls_files.sort.reverse : filtered_ls_files.sort
+  end
+
+  def show_files_by_long_format(ls_files)
+    max_length = bring_file_max_length(ls_files)
+    puts "total #{ls_files.sum(&:block_size)}"
+    ls_files.each do |ls_file|
+      time_format = bring_time_format(ls_file.file_stat)
+      ls_file_properties = []
+      ls_file_properties << ls_file.mode
+      ls_file_properties << ls_file.hard_links.to_s.rjust(max_length[:hard_links])
+      ls_file_properties << ls_file.owner_name.to_s.rjust(max_length[:owner_name])
+      ls_file_properties << ls_file.owner_group_name.to_s.rjust(max_length[:owner_group_name])
+      ls_file_properties << ls_file.size_or_device_info.to_s.rjust(max_length[:size])
+      ls_file_properties << ls_file.last_modified_time.strftime(time_format)
+      ls_file_properties << ls_file.name
+      puts ls_file_properties.join(' ')
+    end
+  end
+
+  def bring_file_max_length(ls_files)
+    {
+      hard_links: ls_files.map { |ls_file| ls_file.hard_links.to_s.length }.max,
+      owner_name: ls_files.map { |ls_file| ls_file.owner_name.to_s.length }.max,
+      owner_group_name: ls_files.map { |ls_file| ls_file.owner_group_name.to_s.length }.max,
+      size: ls_files.map { |ls_file| ls_file.size_or_device_info.to_s.length }.max
+    }
+  end
+
+  def bring_time_format(file_stat)
+    if Time.now.year == file_stat.mtime.year
+      '%b %e %H:%M'
+    else
+      '%b %e  %Y'
+    end
+  end
+
+  def show_files_by_default_format(ls_files)
+    rows = (ls_files.size.to_f / COLS).ceil
+    nested_files = Array.new(rows) { Array.new(COLS) }
+    ls_files.each_with_index { |ls_file, index| nested_files[index % rows][index / rows] = ls_file }
+    max_lengths = bring_file_name_max_lengths(nested_files)
+    nested_files.each do |row_files|
+      formatted_row_files = row_files.compact.map.with_index do |file, index|
+        file.name.ljust(max_lengths[index] + 2)
+      end
+      puts formatted_row_files.join
+    end
+  end
+
+  def bring_file_name_max_lengths(nested_files)
+    max_lengths = [0] * COLS
```

Comment:
```suggestion
    max_lengths = Array.new(COLS, 0)
```

こっちの方がコードの意図が伝わりやすいかな